### ISSUE TITLE
Passing "sub Stores" for subscription

### ISSUE THREAD

**Issue by raheelahmad** [ReSwift/ReSwift#69]
This comes up when passing down stores as dependencies.

Currently being able to subscribe for a sub-state is awesome, as the `newState`  implementation is filtered down.
It would be nice to bring the same `State` filtering to `Store`s.

In other words, be able to generate sub-stores from the main store (just as a lens; there will still be a single concrete store).
Then we can pass the sub-store so that subselection is even more filtered.

As an example:

```swift
struct InstrumentState: StateType {
    let allInstruments: [Instrument]
    let currentIndex: Int?
}

protocol HasInstrumentState {
  var instrumentsState: InstrumentsState { get }
}

struct AppState: StateType, HasInstrumentState {
  let instrumentState: InstrumentState
}


class SmallSubscriber: StoreSubscriber {

  // --- Subscriber can then just expect a "smaller" Store
  init(store: Store<InstrumentState>) {
      store.subscribe(self) {
          return $0.current
      }
  }
}

// But if we start with appState
let instrumentsState =  InstrumentState(allInstruments: [...], currentIndex: 0)
let appState = AppState(instrumentState: instrumentsState)
let mainStore = Store<AppState>(reducer: MyReducer(), state: appState)

// then we can't pass that store to our subscriber:
let subscriber = SmallSubscriber(store: mainStore)
```

- we then expose minimal information
- better at scaling: large apps may have deep nesting of states in the store. Being able to pass a smaller store of a deeper nesting would be great

Maybe this is already possible?

**Comment by Aleksion** [comment]
I don't think it is at the moment. But I rather like the idea. If we can figure out how to do it gracefully :)... 
I haven't had enough time to play around with the newState filtering @Ben-G implemented.

I suppose this would also make it easier to import external subscribers, since they simply specify that they need a Store which state is <InstrumentState>. 

Do you have any suggestions on how to implement this?

**Comment by raheelahmad** [comment]
I am not sure what a good "Swifty" design would be, yet.

Looking at Om, the analog would the concept of ["cursor"](https://github.com/omcljs/om/wiki/Cursors).

**Comment by Ben-G** [comment]
Substate selection is now solved by #61.
